src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs

```diff
@@ -0,0 +1,112 @@
+
+        public static event NetworkAddressChangedEventHandler NetworkAddressChanged
+        {
+            add
+            {
+                lock (s_lockObj)
+                {
+                    if (s_socket == 0)
+                    {
+                        CreateSocket();
+                    }
+
+                    s_addressChangedSubscribers += value;
+                }
+            }
+            remove
+            {
+                lock (s_lockObj)
+                {
+                    if (s_addressChangedSubscribers == null)
+                    {
+                        Debug.Assert(s_socket == 0);
+                        return;
+                    }
+
+                    s_addressChangedSubscribers -= value;
+                    if (s_addressChangedSubscribers == null)
+                    {
+                        CloseSocket();
+                    }
+                }
+            }
+        }
+
+        private static void CreateSocket()
+        {
+            Debug.Assert(s_socket == 0);
+            int newSocket;
+            Interop.Error result = Interop.Sys.CreateNetworkChangeListenerSocket(out newSocket);
+            if (result != Interop.Error.SUCCESS)
+            {
+                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
+                throw new NetworkInformationException(message);
+            }
+
+            s_socket = newSocket;
+            Task.Run(() => LoopReadSocket(s_socket));
+        }
+
+        private static void CloseSocket()
+        {
+            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket);
+            if (result != Interop.Error.SUCCESS)
+            {
+                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
+                throw new NetworkInformationException(message);
+            }
+        }
+
+        private static void LoopReadSocket(int socket)
+        {
+            while (socket == s_socket)
+            {
+                Interop.Sys.NetworkChangeKind kind = Interop.Sys.ReadSingleEvent(s_socket);
+                if (kind == Interop.Sys.NetworkChangeKind.None)
+                {
+                    continue;
+                }
+                else
+                {
+                    lock (s_lockObj)
+                    {
+                        if (socket == s_socket)
+                        {
+                            OnSocketEvent(kind);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static void OnSocketEvent(Interop.Sys.NetworkChangeKind kind)
+        {
+            switch (kind)
+            {
+                case Interop.Sys.NetworkChangeKind.AddressAdded:
+                case Interop.Sys.NetworkChangeKind.AddressRemoved:
+                    if (s_addressChangedSubscribers != null)
+                    {
+                        s_addressChangedSubscribers(null, EventArgs.Empty);
```

Comment:
This code assumes that the lock is being held; otherwise s_addressChangeSubscribers could go from non-null to null between the check and the invocation.  If you decide it's ok to do this while holding the lock, we should add a Debug.Assert for Monitor.IsEntered(s_lockObj).  If not, this should be changed to store the delegate into a local, and then do the check and invoke on that local.